File: Medium/167_Two_Sum_II_Input_Array_Is_Sorted/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int left = 0;
        int right = numbers.size() - 1;

        while (left < right) {
            int sum = numbers[left] + numbers[right];

            if (sum == target) {
                return {left + 1, right + 1};
            } else if (sum < target) {
                left++;
            } else {
                right--;
            }
        }

        return {};
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    vector<int> numbers1 = {2,7,11,15};
    int target1 = 9;
    printVec(solution.twoSum(numbers1, target1));

    // test cases 2
    vector<int> numbers2 = {2,3,4};
    int target2 = 6;
    printVec(solution.twoSum(numbers2, target2));

    // test cases 3
    vector<int> numbers3 = {-1,0};
    int target3 = -1;
    printVec(solution.twoSum(numbers3, target3));
    
    return 0;
}

File: Medium/167_Two_Sum_II_Input_Array_Is_Sorted/Solution.go
package main

import "fmt"

// Past the function from leetcode here
func twoSum(numbers []int, target int) []int {
	left := 0
	right := len(numbers) - 1

	for left < right {
		sum := numbers[left] + numbers[right]

		if sum == target {
			return []int{left + 1, right + 1}
		} else if sum < target {
			left++
		} else {
			right--
		}
	}

	return []int{}
}

func main() {
	// test cases 1
	numbers1 := []int{2, 7, 11, 15}
	target1 := 9
	fmt.Println(twoSum(numbers1, target1))

	// test cases 2
	numbers2 := []int{2, 3, 4}
	target2 := 6
	fmt.Println(twoSum(numbers2, target2))

	// test cases 3
	numbers3 := []int{-1, 0}
	target3 := -1
	fmt.Println(twoSum(numbers3, target3))
}
